### DISCUSSION TITLE
Virtual Coffee's Technical Task Challenges

### DISCUSSION THREAD

**Discussion by Arvind644** [Virtual-Coffee/virtualcoffee.io#1140] [New Ideas]
Please read and respond with your thoughts, questions, and observations regarding the following proposal:

# Virtual Coffee's Technical Task Challenges

## Summary

Virtual Coffee's Technical Task Challenges are created/aggregated by VC members to help other VC members gain practical experience in how to use technical skills in real-world scenarios.

## Benefits
These task challenges will greatly benefit the entire Virtual Coffee community. They will serve our members who are current working professionals by allowing them to up-skill in select technologies and tools. They will also help our job seekers, especially those with no prior professional experience in the tech industry. The unique added value of the technical task challenges is that they're modeled after real-world scenarios, lending these exercises a quality that far surpasses what's readily available on sites like LeetCode and Codewars.

## Logistics

We will create a repository to house the challenges and related information. The repo will contain…

- a `README` file which explains what the repo contains and how to use it
- a directory-type document where VC members can post their current and previous job titles, then list routine tasks they perform(ed) for that job
- a new folder per challenge, each with its own `README` file (we will provide an example for others to follow)

### Adding Your Experience

As a VC member with real-world experience, you may want to list some of your routine technical tasks. To do this, first create a fork of the repository and a new branch titled `add-to-directory`. Add your name, job title(s), years in the role(s), and tasks to the directory document. Then, make a pull request back to the main project.

In doing this, you are signaling to other members that you are open for conversation about the listed tasks. If you so choose, you can post a challenge (or challenges) related to one or more of your listed tasks.

**NOTE:** You **should not** post any proprietary data or company-specific tasks. Focus only on transferable tools and technology.

### Posting a New Technical Task Challenge

As a VC member with real-world experience, you may want to post a technical task challenge. To do this, create a fork of the repository with a new branch titled `add-<challenge-name>-challenge`. Begin by creating a new folder to hold your challenge, then add a `README` file. Link from the relevant task in the directory document to the `README` file in your task challenge folder.

This `README` should explain the goal(s) of the task challenge and how to get started. It should also list…

- the VC member who posted the task challenge (you!)
- the challenge source/author
- the skill level this challenge is targeting
- any VC members currently working on the task challenge
- any VC members who have successfully completed the task challenge, including a link to their fork of the repo

**NOTE:** You do not have to create the task challenge from scratch, however, you **must have completed it before posting it** as you will be responsible for giving final feedback to VC members who attempt your challenge.

#### Technical Task Challenge Skill Levels

We will divide task challenges into beginner, intermediate, and advanced levels. Challenges can be multi-part, increasing in difficulty the further along you go.

- **Beginner level tasks:** for a person learning to code (<2 years of experience)
- **Intermediate level tasks:** with a focus on the working professional and building up skills (2-5 years of experience)
- **Advanced level tasks:** for seniors who want a challenge (5+ years of experience)

### Working on a Technical Task Challenge

As a VC member looking to build your skills, you may want to try several technical task challenges. To work on a task challenge, create your own fork of the repository and a new branch titled `add-challenge-participant`. Type your name in the appropriate spot of the `README` where it lists the task challenge participants, and send a pull request back to the main repo.

#### Who Do You Go To for Help?

There are several avenues of help available to VC members who take on a task challenge. The following in the order in which you should seek out help:

1. **If there are other VC members currently attempting the task challenge**, partner up with them on any help you need or any questions you may have.
2. **If any VC members have successfully completed the challenge**, review supplemental resources they created after completing the task challenge (if any) and reach out to them with questions.
3. **If you are the first person to attempt the task challenge**, communicate any questions you have to the VC member who posted the challenge and reach out to them directly for help.

#### Submitting Your Technical Task Challenge

Unless you are the first person to attempt the task challenge, you will be required to have another VC member peer review your work before submitting your final solution to the challenge poster. If there are other VC members currently attempting the task challenge, you should pair up and peer review each other's work. If there are no other members currently attempting the task challenge, ask a member who has successfully completed it to review your work.

After your solution has been peer reviewed and necessary changes made, you may proceed to submit it to the challenge originator for final analysis and feedback. Once you receive your feedback from the challenge poster and are told you've successfully completed it, you may create a new pull request to move your name in the technical task challenge `README` from the list of participants to the list of successful completions. Link back to your challenge solution in your fork if you'd like!

## Example

At work, one of my tasks is to implement a CI/CD setup for a React/Node application using GitLab CI and Jenkins. If I wanted to create a task challenge for this, I would fork the repository, add my information to the directory document, and create a new folder called `Implement CI/CD Setup for a React/Node Application Using GitLab CI and Jenkins`. I would add a `README` file with an overview of the challenge and instructions on how to proceed and link that back to the directory document.

Since I've done this before and am very comfortable with the process, I can help whoever tries it out first. Then as more people attempt the challenge, the pool of people who can help answer questions and debug grows, so I don't become overwhelmed with requests for support.

## Going Further

There are a few more ideas for what we can do after a VC member has completed a task challenge. These are completely optional and open to change.

1. The challenge originator can provide them with a list of resources and/or topics to follow up on and expand their newfound knowledge.
2. The challenge completer can create a resource of their own covering what they just learned. Options include but are not limited to: blog posts, videos/live streams, Twitter spaces, social media posts, podcasts, etc. We would want to link them in the repository, but they can be shared outside the community as well.

**Comment by juliaseid** [comment]
I like this idea a lot! I'm not sure how well my day-to-day work translates into this kind of challenge, but once a few people have created some examples, I'll see if I can follow their lead. 

I'd also be game to collaborate with someone. My approximate area of expertise is navigating a new-to-me, poorly documented dev tool and creating a demo app and how-to guide, often in 2-3 different languages/frameworks. I do it as a consultant to the API companies when they're building out their docs sites, but it's similar to what folks might do internally when considering adding a new tool for their team. It might be useful to get ideas from others about when you've gone through a process like that internally so I can target a couple of specific use cases.

**Reply by Arvind644** [reply]
@juliaseid  you can create tasks related to blogging or writing articles for a specific topic or any research areas. That would be fun and interesting for us also.

**Comment by jmslynn** [comment]
I'm really excited about this. Very creative. I can't wait to see all the contributions.

**Reply by Arvind644** [reply]
@jmslynn Thanks a lot. I am glad that you like my initiative.

**Comment by Narigo** [comment]
I've created a challenge for hiring like "redo our team page and add yourself on it" and provided a really badly styled page with our portraits and bios inconsistently spaced, etc. We wanted to find frontend people with an eye for design. I think adding yourself to the VC members page is already something like this... 😅

Anyways, I could think of a few challenges, I guess. I think the time is the limiting factor though...

**Reply by Arvind644** [reply]
@Narigo its okay, If you create one challenge in a month, it will works also. As it will take people 1-2 weeks to create the solution and peer review and discussion also takes time. Please let me know if you have any other advices for me.

**Comment by tomcudd** [comment]
Great idea, and I'll be contributing to create or assist with creating challenges. A few suggestions when building this idea out. Don't attribute years to level of experience as this can filter some people out who may feel like they're not a good fit. Let people select into what they are comfortable with. Also, anywhere you want to put "coding", try to keep it generic. There's ranges of technical skills that challenges would work for that are not directly writing software. The use cases I would come up with are cloud, system, DevOps, or other related real work type examples that aren't writing software.

**Reply by Arvind644** [reply]
@tomcudd I thought of experience level so that we can divide challenges into three tracks and I want senior developer to try challenges so I put the advanced level tasks as otherwise a senior developer can read that and thought that its easy, I can do that. But you are right, that it will avoid lot of other people to try the challenges. Instead we can add the prerequisite section in it. 
I loved devops and cloud, as I mostly do it in my work. So I am excited for your challenges now. 
Can you start creating your first challenge, so that others will see and it will motivate them to create their own challenges also. Thanks a lot for your advices.

**Reply by tomcudd** [reply]
I want you to keep the 3 tracks, just don't put years by them. Let people figure out on their own where they want to try out. I'll start working on something by the end of the week, but it might be a few days.

**Reply by Arvind644** [reply]
Thanks. I will create a GitHub repo and share it with you.

**Comment by danieltott** [comment]
@Arvind644 this is a great idea!

I think it'd be good for the repo to be under this Org - I would love to create one for you and make you admin to get us rolling. What's a good name for the repo?

**Reply by Arvind644** [reply]
@danieltott can you create a repo named **Technical Task Challenges**. Since it will contain development, devops and documentation challenges, I am not using coding challenges. Thanks a lot.

**Reply by danieltott** [reply]
@Arvind644 done! https://github.com/Virtual-Coffee/Technical-Task-Challenges

**Reply by Arvind644** [reply]
@danieltott Thanks a lot. I will start working on the issue that you have created.

**Reply by danieltott** [reply]
Awesome! Feel free to reach out any time with questions etc 🙌🏼

**Reply by meg-gutshall** [reply]
Thank you @danieltott! We've started moving on it and issues are up. 👍

**Comment by meg-gutshall** [comment]
Hey all! This announcement is going out tomorrow, but @Arvind644 and I wanted the original supporters to hear it here first:

---

Hey team! We are thrilled to announce our newest endeavor: [**Virtual Coffee’s Task Challenges!**](https://github.com/Virtual-Coffee/task-challenges/tree/main) This is your chance to showcase your coding skills, learn from each other, and have fun along the way. Here are the details:

**Overview**

Task Challenges are created by VC members for VC members to gain practical experience based on real-world scenarios.

- **Experienced members:** Share your expertise by adding your professional experience to the [Member Experience Directory](https://github.com/Virtual-Coffee/task-challenges/blob/main/member-experience.md) along with a list of common tasks performed for each position you’ve held. Design a new challenge based on a task you’ve listed so members can acquire the skills needed for their target positions.
- **Skill builders:** Enhance your skill set by completing hands-on tasks and peer reviewing other members’ work.

**How to Participate**

Our first challenge is the Simple Backend API Challenge. Clone the GitHub repo, add your name to the Members Attempting section, and raise a PR along with your coding solution. [Link to Challenge](https://github.com/Virtual-Coffee/task-challenges/blob/main/task-challenges/_example-task-challenge.md)

**Benefits of Participating**

- **For challenge posters:** Guide fellow members and contribute to a supportive learning environment.
- **For challenge participants:** Gain real-world experience, build your portfolio, and receive feedback from professionals.

Unlike abstract problem-solving platforms like LeetCode and Codewars, these challenges are grounded in real-life applications, making them immensely beneficial for professional development and applied learning. In developing new and varied Task Challenges, we’ll be building a repository of valuable resources while fostering collaborative learning within the Virtual Coffee community.

Questions? Reach out to @Arvind644 or @meg-gutshall. We can’t wait to see what you create! Happy coding! 🚀 

---

As always, let us know your thoughts!!
